portal/index.php

```diff
@@ -27,15 +27,52 @@
 
 //includes
 require_once('../interface/globals.php');
+require_once(dirname(__FILE__) . "/lib/appsql.class.php");
+$logit = new ApplicationTable();
 
+use OpenEMR\Common\Crypto\CryptoGen;
 use OpenEMR\Core\Header;
 
 //exit if portal is turned off
 if (!(isset($GLOBALS['portal_onsite_two_enable'])) || !($GLOBALS['portal_onsite_two_enable'])) {
     echo xlt('Patient Portal is turned off');
     exit;
 }
-
+$auth['portal_pwd'] = '';
+if (isset($_GET['woops'])) {
+    unset($_GET['woops']);
+    unset($_SESSION['password_update']);
+}
+if (isset($_GET['forward'])) {
+    $auth = false;
+    if (strlen($_GET['forward']) >= 64) {
+        $crypto = new CryptoGen();
+        $one_time = $crypto->decryptStandard($_GET['forward'], '', 'database');
```

Comment:
for me:
use key from drive and do check to standard check to ensure token not empty/false
(also run cryptCheckStandard() on it first)
(and will set 6 as the lowest decrypt version on the updated decryptStandard() function)

Reply:
is encrypting a string results repeatable ie encrypt the same 32 byte token twice? I notice with database key it's not but how about file?

Reply:
No, this is because encryption does something similar to a salt for hashes. The encryption randomly creates an iv (akin to a salt):
https://github.com/openemr/openemr/blob/master/src/Common/Crypto/CryptoGen.php#L162
which is used in the encryption:
https://github.com/openemr/openemr/blob/master/src/Common/Crypto/CryptoGen.php#L173
and then bundled with the encryption:
https://github.com/openemr/openemr/blob/master/src/Common/Crypto/CryptoGen.php#L184

Reply:
fyi, a full night of reading on iv :)
https://en.wikipedia.org/wiki/Initialization_vector

Reply:
Sorry, not that dedicated. I'm testing changes/move to onetime now and changed key to file for testing. I'll push and let you fixup encrypt stuff.